File: node/tests/pool.test.js
/**
 * pool.test.js — Connection pool tests for VedaDB Node.js driver
 */

class ConnectionPool {
  constructor(factory, options = {}) {
    this.factory = factory;
    this.maxSize = options.maxSize || 10;
    this.maxIdle = options.maxIdle || 5;
    this.waitTimeout = options.waitTimeout || 5000;
    this.available = [];
    this.allConnections = [];
    this.activeCount = 0;
    this.totalCreated = 0;
    this.closed = false;
    this.acquiring = [];
  }

  async acquire() {
    if (this.closed) throw new Error('Pool is closed');

    // Try to get from pool
    if (this.available.length > 0) {
      const conn = this.available.pop();
      conn.inUse = true;
      return conn;
    }

    // Create new if under max
    if (this.totalCreated < this.maxSize) {
      this.totalCreated++;
      const raw = this.factory();
      const conn = new PooledConnection(raw, this.totalCreated, this);
      conn.inUse = true;
      this.allConnections.push(conn);
      this.activeCount++;
      return conn;
    }

    // Wait for available connection
    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        reject(new Error('Pool exhausted: wait timeout'));
      }, this.waitTimeout);

      const check = () => {
        if (this.available.length > 0) {
          clearTimeout(timer);
          const conn = this.available.pop();
          conn.inUse = true;
          resolve(conn);
        } else {
          setTimeout(check, 10);
        }
      };
      setTimeout(check, 10);
    });
  }

  release(conn) {
    conn.inUse = false;
    this.activeCount--;
    if (this.available.length < this.maxIdle) {
      this.available.push(conn);
    }
  }

  close() {
    this.closed = true;
    for (const conn of this.allConnections) {
      if (conn.close) conn.close();
    }
    this.available = [];
  }
}

class PooledConnection {
  constructor(connection, id, pool) {
    this.connection = connection;
    this.id = id;
    this.pool = pool;
    this.inUse = false;
    this.createdAt = Date.now();
  }

  release() {
    this.pool.release(this);
  }

  get isValid() {
    return !this.connection.closed;
  }
}

describe('ConnectionPool', () => {
  const mockFactory = () => ({ closed: false, query: () => Promise.resolve([]) });

  describe('Acquire', () => {
    test('should acquire new connection', async () => {
      const pool = new ConnectionPool(mockFactory, { maxSize: 5 });
      const conn = await pool.acquire();
      expect(conn).toBeDefined();
      expect(conn.inUse).toBe(true);
      conn.release();
      pool.close();
    });

    test('should reuse released connection', async () => {
      const pool = new ConnectionPool(mockFactory, { maxSize: 5 });
      const conn1 = await pool.acquire();
      const id1 = conn1.id;
      conn1.release();
      const conn2 = await pool.acquire();
      expect(conn2.id).toBe(id1);
      conn2.release();
      pool.close();
    });

    test('should track total created', async () => {
      const pool = new ConnectionPool(mockFactory, { maxSize: 5 });
      expect(pool.totalCreated).toBe(0);
      const conn = await pool.acquire();
      expect(pool.totalCreated).toBe(1);
      conn.release();
      pool.close();
    });
  });

  describe('Pool Exhaustion', () => {
    test('should timeout when exhausted', async () => {
      const pool = new ConnectionPool(mockFactory, { maxSize: 1, waitTimeout: 50 });
      const conn = await pool.acquire();
      await expect(pool.acquire()).rejects.toThrow('timeout');
      conn.release();
      pool.close();
    });

    test('should enforce max connections', async () => {
      const pool = new ConnectionPool(mockFactory, { maxSize: 3 });
      const conns = [];
      for (let i = 0; i < 3; i++) {
        conns.push(await pool.acquire());
      }
      expect(pool.totalCreated).toBe(3);
      for (const c of conns) c.release();
      pool.close();
    });
  });

  describe('Release', () => {
    test('should return connection to pool', async () => {
      const pool = new ConnectionPool(mockFactory, { maxSize: 5 });
      const conn = await pool.acquire();
      conn.release();
      const conn2 = await pool.acquire();
      expect(conn2).toBeDefined();
      conn2.release();
      pool.close();
    });
  });

  describe('Close', () => {
    test('should close pool', () => {
      const pool = new ConnectionPool(mockFactory, { maxSize: 5 });
      pool.close();
      expect(pool.closed).toBe(true);
    });

    test('should reject acquire after close', async () => {
      const pool = new ConnectionPool(mockFactory, { maxSize: 5 });
      pool.close();
      await expect(pool.acquire()).rejects.toThrow('closed');
    });
  });

  describe('Concurrency', () => {
    test('should handle concurrent acquire/release', async () => {
      const pool = new ConnectionPool(mockFactory, { maxSize: 5, waitTimeout: 1000 });
      const promises = [];

      for (let i = 0; i < 20; i++) {
        promises.push((async () => {
          const conn = await pool.acquire();
          await new Promise(r => setTimeout(r, 1));
          conn.release();
        })());
      }

      await Promise.all(promises);
      pool.close();
    });

    test('should handle stress test', async () => {
      const pool = new ConnectionPool(mockFactory, { maxSize: 5, waitTimeout: 2000 });
      const promises = [];
      let acquired = 0;

      for (let i = 0; i < 50; i++) {
        promises.push((async () => {
          try {
            const conn = await pool.acquire();
            acquired++;
            await new Promise(r => setTimeout(r, 1));
            conn.release();
          } catch (e) {
            // Expected for some
          }
        })());
      }

      await Promise.all(promises);
      expect(acquired).toBeGreaterThan(0);
      pool.close();
    });
  });
});

module.exports = { ConnectionPool, PooledConnection };

File: go/vedadb/pool_test.go
// pool_test.go — Connection pool tests for VedaDB Go driver
package vedadb

import (
	"context"
	"errors"
	"sync"
	"sync/atomic"
	"testing"
	"time"
)

// PoolConn represents a pooled connection
type PoolConn struct {
	id       int64
	client   *Client
	pool     *Pool
	created  time.Time
	inUse    bool
}

func (pc *PoolConn) Release() {
	pc.pool.release(pc)
}

func (pc *PoolConn) IsValid() bool {
	return !pc.client.closed
}

// Pool manages a pool of reusable connections
type Pool struct {
	mu          sync.Mutex
	endpoint    string
	maxConns    int
	maxIdle     int
	connections []*PoolConn
	available   chan *PoolConn
	activeCount int64
	totalCreated int64
	waitTimeout time.Duration
	factory     func() (*Client, error)
	closed      bool
}

func NewPool(endpoint string, maxConns, maxIdle int, waitTimeout time.Duration) *Pool {
	return &Pool{
		endpoint:    endpoint,
		maxConns:    maxConns,
		maxIdle:     maxIdle,
		connections: make([]*PoolConn, 0, maxConns),
		available:   make(chan *PoolConn, maxConns),
		waitTimeout: waitTimeout,
		factory: func() (*Client, error) {
			return Connect(context.Background(), endpoint)
		},
	}
}

func (p *Pool) Acquire(ctx context.Context) (*PoolConn, error) {
	p.mu.Lock()
	if p.closed {
		p.mu.Unlock()
		return nil, errors.New("pool is closed")
	}

	// Try to get from available
	select {
	case conn := <-p.available:
		p.mu.Unlock()
		conn.inUse = true
		atomic.AddInt64(&p.activeCount, 1)
		return conn, nil
	default:
	}

	// Create new if under max
	if len(p.connections) < p.maxConns {
		atomic.AddInt64(&p.totalCreated, 1)
		client, err := p.factory()
		if err != nil {
			p.mu.Unlock()
			return nil, err
		}
		conn := &PoolConn{
			id:      p.totalCreated,
			client:  client,
			pool:    p,
			created: time.Now(),
			inUse:   true,
		}
		p.connections = append(p.connections, conn)
		atomic.AddInt64(&p.activeCount, 1)
		p.mu.Unlock()
		return conn, nil
	}
	p.mu.Unlock()

	// Wait for available connection
	select {
	case conn := <-p.available:
		conn.inUse = true
		atomic.AddInt64(&p.activeCount, 1)
		return conn, nil
	case <-ctx.Done():
		return nil, ctx.Err()
	case <-time.After(p.waitTimeout):
		return nil, errors.New("pool exhausted: wait timeout")
	}
}

func (p *Pool) release(conn *PoolConn) {
	conn.inUse = false
	atomic.AddInt64(&p.activeCount, -1)
	select {
	case p.available <- conn:
	default:
		// Pool full, discard
	}
}

func (p *Pool) ActiveCount() int {
	return int(atomic.LoadInt64(&p.activeCount))
}

func (p *Pool) TotalCreated() int64 {
	return atomic.LoadInt64(&p.totalCreated)
}

func (p *Pool) Close() error {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.closed = true
	for _, conn := range p.connections {
		if conn.client != nil {
			conn.client.Close()
		}
	}
	return nil
}

func TestPoolAcquire(t *testing.T) {
	t.Run("acquire_new_connection", func(t *testing.T) {
		p := NewPool("mock://test", 10, 5, 1*time.Second)
		defer p.Close()

		conn, err := p.Acquire(context.Background())
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if conn == nil {
			t.Fatal("expected connection, got nil")
		}
		if conn.pool != p {
			t.Error("expected connection to belong to pool")
		}
		conn.Release()
	})

	t.Run("acquire_reuses_released", func(t *testing.T) {
		p := NewPool("mock://test", 10, 5, 1*time.Second)
		defer p.Close()

		conn1, _ := p.Acquire(context.Background())
		conn1.Release()

		conn2, _ := p.Acquire(context.Background())
		defer conn2.Release()

		if conn1.id != conn2.id {
			t.Error("expected same connection to be reused")
		}
	})

	t.Run("acquire_tracks_active", func(t *testing.T) {
		p := NewPool("mock://test", 10, 5, 1*time.Second)
		defer p.Close()

		if p.ActiveCount() != 0 {
			t.Errorf("expected 0 active initially, got %d", p.ActiveCount())
		}

		conn, _ := p.Acquire(context.Background())
		if p.ActiveCount() != 1 {
			t.Errorf("expected 1 active, got %d", p.ActiveCount())
		}

		conn.Release()
		// Allow a moment for the release to process
		time.Sleep(10 * time.Millisecond)
	})
}

func TestPoolRelease(t *testing.T) {
	t.Run("release_returns_to_pool", func(t *testing.T) {
		p := NewPool("mock://test", 10, 5, 1*time.Second)
		defer p.Close()

		conn, _ := p.Acquire(context.Background())
		conn.Release()

		// Should be able to acquire again
		conn2, err := p.Acquire(context.Background())
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		conn2.Release()
	})

	t.Run("release_invalid_connection", func(t *testing.T) {
		p := NewPool("mock://test", 10, 5, 1*time.Second)
		defer p.Close()

		conn, _ := p.Acquire(context.Background())
		conn.client.closed = true
		conn.Release()
	})
}

func TestPoolExhausted(t *testing.T) {
	t.Run("wait_timeout_when_exhausted", func(t *testing.T) {
		p := NewPool("mock://test", 1, 1, 50*time.Millisecond)
		defer p.Close()

		conn, _ := p.Acquire(context.Background())
		defer conn.Release()

		ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
		defer cancel()

		_, err := p.Acquire(ctx)
		if err == nil {
			t.Fatal("expected error when pool exhausted")
		}
	})

	t.Run("context_cancel_while_waiting", func(t *testing.T) {
		p := NewPool("mock://test", 1, 1, 1*time.Hour)
		defer p.Close()

		conn, _ := p.Acquire(context.Background())
		defer conn.Release()

		ctx, cancel := context.WithCancel(context.Background())
		go func() {
			time.Sleep(50 * time.Millisecond)
			cancel()
		}()

		_, err := p.Acquire(ctx)
		if err == nil {
			t.Fatal("expected error after context cancel")
		}
	})

	t.Run("max_connections_enforced", func(t *testing.T) {
		created := int64(0)
		p := NewPool("mock://test", 3, 3, 100*time.Millisecond)
		p.factory = func() (*Client, error) {
			atomic.AddInt64(&created, 1)
			return &Client{endpoint: "mock"}, nil
		}
		defer p.Close()

		conns := make([]*PoolConn, 3)
		for i := 0; i < 3; i++ {
			conns[i], _ = p.Acquire(context.Background())
		}

		if created != 3 {
			t.Errorf("expected 3 created, got %d", created)
		}

		for _, c := range conns {
			c.Release()
		}
	})
}

func TestPoolConcurrent(t *testing.T) {
	t.Run("concurrent_acquire_release", func(t *testing.T) {
		p := NewPool("mock://test", 10, 10, 1*time.Second)
		p.factory = func() (*Client, error) {
			return &Client{endpoint: "mock"}, nil
		}
		defer p.Close()

		var wg sync.WaitGroup
		for i := 0; i < 50; i++ {
			wg.Add(1)
			go func() {
				defer wg.Done()
				conn, err := p.Acquire(context.Background())
				if err != nil {
					t.Errorf("unexpected error: %v", err)
					return
				}
				time.Sleep(1 * time.Millisecond)
				conn.Release()
			}()
		}
		wg.Wait()
	})

	t.Run("concurrent_stress_test", func(t *testing.T) {
		p := NewPool("mock://test", 5, 5, 1*time.Second)
		p.factory = func() (*Client, error) {
			return &Client{endpoint: "mock"}, nil
		}
		defer p.Close()

		var wg sync.WaitGroup
		errCount := int64(0)
		successCount := int64(0)

		for i := 0; i < 100; i++ {
			wg.Add(1)
			go func() {
				defer wg.Done()
				conn, err := p.Acquire(context.Background())
				if err != nil {
					atomic.AddInt64(&errCount, 1)
					return
				}
				atomic.AddInt64(&successCount, 1)
				time.Sleep(1 * time.Millisecond)
				conn.Release()
			}()
		}
		wg.Wait()

		if successCount == 0 {
			t.Error("expected some successful acquisitions")
		}
	})
}

func TestPoolClose(t *testing.T) {
	t.Run("close_rejects_new_acquire", func(t *testing.T) {
		p := NewPool("mock://test", 10, 5, 1*time.Second)
		p.Close()

		_, err := p.Acquire(context.Background())
		if err == nil {
			t.Fatal("expected error after pool close")
		}
	})

	t.Run("close_is_idempotent", func(t *testing.T) {
		p := NewPool("mock://test", 10, 5, 1*time.Second)
		p.Close()
		err := p.Close()
		if err != nil {
			t.Fatalf("close should be idempotent: %v", err)
		}
	})
}
